### ISSUE TITLE
Too many ACL cause database errors

### ISSUE THREAD

**Issue by mgiacomoli** [mediawiki4intranet/IntraACL#49]
Hello,

Mediawiki: 1.23.13
IntraACL: 2.1.8 (4d877b27c256f59fa53ba732759508f26b4c3fae)

Side note:
I wrote a maintenance script that create page ACL for a group of pages, basically it creates a WikiPage with the rules you can see in IntraACL Editor as page text, and "ACL:Page/<page_name>" as title. The rules are always the same for each page. I hope this is the correct way to add acls programmatically. 

Issue:
When I tried the script for a few pages it worked and I got all the ACL set for that pages. Then I tried the script with a big set of pages (~9000), and it returned without errors. But then every wiki page I try to open I get nothing but this:

```
[dc94d55d] /index.php/Wikipedia Exception from line 2030 of /<mediawiki_path>/mediawiki/includes/db/Database.php: DatabaseBase::makeList: empty input for field pe_id

Backtrace:

#0 /<mediawiki_path>/mediawiki/includes/db/Database.php(1640): DatabaseBase->makeList(array, integer)
#1 /<mediawiki_path>/mediawiki/includes/db/Database.php(1588): DatabaseBase->selectSQLText(string, string, array, string, array, array)
#2 /<mediawiki_path>/mediawiki/extensions/IntraACL/storage/SQL_SD.php(32): DatabaseBase->select(string, string, array, string, array)
#3 /<mediawiki_path>/mediawiki/extensions/IntraACL/includes/Definition.php(534): IntraACL_SQL_SD->getRules(array, array)
#4 /<mediawiki_path>/mediawiki/extensions/IntraACL/includes/Evaluator.php(413): IACLDefinition::userCan(integer, integer, array, integer)
#5 /<mediawiki_path>/mediawiki/extensions/IntraACL/includes/Evaluator.php(358): IACLEvaluator::hasSD(Title, integer, integer, integer)
#6 /<mediawiki_path>/mediawiki/extensions/IntraACL/includes/Evaluator.php(143): IACLEvaluator::userCan_Switches(Title, User, string)
#7 [internal function]: IACLEvaluator::userCan(Title, User, string, string)
#8 /<mediawiki_path>/mediawiki/includes/Hooks.php(204): call_user_func_array(string, array)
#9 /<mediawiki_path>/mediawiki/includes/GlobalFunctions.php(4061): Hooks::run(string, array, NULL)
#10 /<mediawiki_path>/mediawiki/includes/Title.php(2044): wfRunHooks(string, array)
#11 /<mediawiki_path>/mediawiki/includes/Title.php(2458): Title->checkPermissionHooks(string, User, array, boolean, boolean)
#12 /<mediawiki_path>/mediawiki/includes/Title.php(1912): Title->getUserPermissionsErrorsInternal(string, User, boolean, boolean)
#13 /<mediawiki_path>/mediawiki/includes/Title.php(5032): Title->userCan(string)
#14 /<mediawiki_path>/mediawiki/includes/Title.php(258): Title->checkAccessControl()
#15 /<mediawiki_path>/mediawiki/includes/Wiki.php(94): Title::newFromURL(string)
#16 /<mediawiki_path>/mediawiki/includes/Wiki.php(142): MediaWiki->parseTitle()
#17 /<mediawiki_path>/mediawiki/includes/Wiki.php(573): MediaWiki->getTitle()
#18 /<mediawiki_path>/mediawiki/includes/Wiki.php(506): MediaWiki->main()
#19 /<mediawiki_path>/mediawiki/index.php(47): MediaWiki->run()
#20 {main}
```
I've done some tests and here are the results/info:

- Number of intraac_rules rows after which the problem is present (note: every acl I create adds 4 rows to this table, so the critical threshold, if any, should be between 2010 and 2013):
```
mysql> select count(*) from intraacl_rules;

+----------+
| count(*) |
+----------+
|     2013 |
+----------+
```
- Type and number of pages in ACL namespace when the problem appeared
-- ACL:Group/ - 3 pages
-- ACL:Namespace/ - 3 pages
-- ACL:Right/ - 2 pages
-- ACL:Page/ - 498 pages
- The problem cause doesn't depend on which page I apply the ACL to, since different sets of pages (with the same number of pages) generate the same problem.
- The problem doesn't depend on the number of page saved on the database, because I added ACLs until count(*) for intraacl_rules was 2009 (one ACL away before the problem), then added 2 standard pages to the wiki, and the wiki was still ok. Then added one more ACL (using my script or IntraACL Editor doesn't matter) and I got the problem.
- Disabling IntraACL extension the wiki works well like before.

- Even loading the page with ?hacllog=true gave me no other infos.

I'm willing to do more tests if needed
Best regards

**Comment by vitalif** [comment]
Can you please add `var_dump($peID);` before line 534 (before `$rules = IACLStorage::get('SD')->getRules($where, $options);`) in includes/Definition.php and post the last value dumped here?

**Comment by mgiacomoli** [comment]
Sure man!
Here is the last value:
```
array (size=0)
  empty
```

**Comment by mgiacomoli** [comment]
Hi!
Can you please provide me any update about the progress of this issue? Is this actually a bug? Do you have some idea about the cause of this bug?
I'm still available for tests ;-)
Best regards

**Comment by vitalif** [comment]
Ok, just fixed it, try the master branch or v2.1.9.1

**Comment by mgiacomoli** [comment]
Well, installed v2.1.9.1 and I don't see this problem anymore. Anyway another problem arose.

In fact after a lot of ACLs are created, NS rules are ignored. For example:
NS Rule:
```
{{#access: assigned to = User:Test | actions = *, manage}}
{{#manage rights: assigned to = User:Admin}}
```
Page Rule:
```
{{#manage rights: assigned to = User:Admin}}
```
hacllog with few and lot of ACLs are attached:
[hacllog-few-ok.txt](https://github.com/mediawiki4intranet/IntraACL/files/193425/hacllog-few-ok.txt)
[hacllog-lot-ko.txt](https://github.com/mediawiki4intranet/IntraACL/files/193426/hacllog-lot-ko.txt)

Best regards

**Comment by vitalif** [comment]
Ok, fixed this part too. Re-update to v2.1.9.1 with git pull --rebase and git pull --tags (I've slightly played around with git commits... :))

**Comment by mgiacomoli** [comment]
Hi!
Downloaded new tar.gz for v2.1.9.1 (opened Definitions.php and the changes are there) and the problem is still not fixed :-(

Attached hacllogs (seems the same as before)
[hacllog-few-ok.txt](https://github.com/mediawiki4intranet/IntraACL/files/193530/hacllog-few-ok.txt)
[hacllog-lot-ko.txt](https://github.com/mediawiki4intranet/IntraACL/files/193531/hacllog-lot-ko.txt)

**Comment by vitalif** [comment]
Interesting, I checked for this bug by lowering $iaclPreloadLimit to 1 (basically "preload (almost) nothing"), and it works with 1 after the fix, but does not work with $iaclPreloadLimit > 1 :):) (i.e. tests/evaluation.php does not pass)
I'll fix it shortly.

**Comment by vitalif** [comment]
Finally it should work without problem. I've rewritten v2.1.9.1 one more time... :)

**Comment by mgiacomoli** [comment]
Hi!
After a quick test it seems to be fixed and I didn't find any regression.
Thank you very much @vitalif !

**Comment by mgiacomoli** [comment]
Hi!
After some more testing I think I found another problem related to this issue.
I set $haclgOpenWikiAccess = false;
Then I tried to set an ACL to Special:Search
```
{{#access: assigned to = # | actions = *}}
{{#manage rights: assigned to = User:Test}}
```
With few ACLs it works, but with lots of (unrelated) ACLs it redirects me to Permission Denied page.
Here are the hacllogs:
[hacllog-few-ok.txt](https://github.com/mediawiki4intranet/IntraACL/files/197173/hacllog-few-ok.txt)
[hacllog-lot-ko.txt](https://github.com/mediawiki4intranet/IntraACL/files/197174/hacllog-lot-ko.txt)

Best regards
